app/src/main/java/org/mozilla/focus/browser/BrowserNavigationOverlay.kt

```diff
@@ -184,4 +261,166 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
+        tiles.add(0, homeTile)
+        notifyItemInserted(0)
+    }
+
+    fun removeTileFromAdapter(url: String) {
+        for ((index, tile) in tiles.withIndex()) {
+            if (tile.url == url) {
+                removeItemAtPosition(index)
+                break
+            }
+        }
+    }
+
+    fun removeItemAtPosition(position: Int) {
+        if (position > -1 && position < itemCount) {
+            tiles.removeAt(position)
+            notifyItemRemoved(position)
+        }
+    }
+
+    override fun getItemCount() = tiles.size
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = TileViewHolder(
+            LayoutInflater.from(parent.context).inflate(R.layout.home_tile, parent, false)
+    )
+}
+
+private fun onBindBundledHomeTile(holder: TileViewHolder, tile: BundledHomeTile) = with (holder) {
+    val bitmap = BundledTilesManager.getInstance(itemView.context).loadImageFromPath(itemView.context, tile.imagePath)
+    iconView.setImageBitmap(bitmap)
+
+    titleView.text = tile.title
+}
+
+private fun onBindCustomHomeTile(uiLifecycleCancelJob: Job, holder: TileViewHolder, item: CustomHomeTile) = with (holder) {
+    launch(uiLifecycleCancelJob + UI, CoroutineStart.UNDISPATCHED) {
+        val validUri = item.url.toJavaURI()
+
+        val screenshotDeferred = async {
+            val homeTileCornerRadius = itemView.resources.getDimension(R.dimen.home_tile_corner_radius)
+            val homeTilePlaceholderCornerRadius = itemView.resources.getDimension(R.dimen.home_tile_placeholder_corner_radius)
+            val screenshot = HomeTileScreenshotStore.read(itemView.context, item.id)?.withRoundedCorners(homeTileCornerRadius)
+            screenshot ?: HomeTilePlaceholderGenerator.generate(itemView.context, item.url).withRoundedCorners(homeTilePlaceholderCornerRadius)
+        }
+
+        val titleDeferred = if (validUri == null) {
+            CompletableDeferred(item.url)
+        } else {
+            async {
+                val subdomainDotDomain = FormattedDomain.format(itemView.context, validUri, false, 1)
+                FormattedDomain.stripCommonPrefixes(subdomainDotDomain)
+            }
+        }
+
+        // We wait for both to complete so we can animate them together.
+        val screenshot = screenshotDeferred.await()
+        val title = titleDeferred.await()
+
+        // NB: Don't suspend after this point (i.e. between view updates like setImage)
+        // so we don't see intermediate view states.
+        // TODO: It'd be less error-prone to launch { /* bg work */ launch(UI) { /* UI work */ } }
+        iconView.setImageBitmap(screenshot)
+        titleView.text = title
+
+        // Animate to avoid pop-in due to thread hand-offs. TODO: animation is janky.
+        AnimatorSet().apply {
+            interpolator = CUSTOM_TILE_ICON_INTERPOLATOR
+            duration = CUSTOM_TILE_TO_SHOW_MILLIS
+
+            val iconAnim = ObjectAnimator.ofInt(iconView, "imageAlpha", 0, 255)
```

Comment:
\[detekt\] INFO: \[style/MagicNumber\] Report magic numbers\. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is\. It's better to declare such numbers as constants and give them a proper name\. By default, \-1, 0, 1, and 2 are not considered to be magic numbers\.